src/rgw/rgw_s3_filter.h

```diff
@@ -14,50 +14,82 @@ struct rgw_s3_key_filter {
   std::string suffix_rule;
   std::string regex_rule;
 
+  bool is_prefix_negative{false};
+  bool is_suffix_negative{false};
+  bool is_regex_negative{false};
+
   bool has_content() const;
 
   void dump(Formatter *f) const;
   bool decode_xml(XMLObj *obj);
   void dump_xml(Formatter *f) const;
 
   void encode(bufferlist& bl) const {
-    ENCODE_START(1, 1, bl);
+    ENCODE_START(2, 1, bl);
       encode(prefix_rule, bl);
       encode(suffix_rule, bl);
       encode(regex_rule, bl);
+      encode(is_prefix_negative, bl);
+      encode(is_suffix_negative, bl);
+      encode(is_regex_negative, bl);
     ENCODE_FINISH(bl);
   }
 
   void decode(bufferlist::const_iterator& bl) {
-    DECODE_START(1, bl);
+    DECODE_START(2, bl);
       decode(prefix_rule, bl);
       decode(suffix_rule, bl);
       decode(regex_rule, bl);
+      if(struct_v >= 2) {
+        decode(is_prefix_negative, bl); 
+        decode(is_suffix_negative, bl); 
+        decode(is_regex_negative, bl);
+      }
     DECODE_FINISH(bl);
   }
 };
 WRITE_CLASS_ENCODER(rgw_s3_key_filter)
 
 using KeyValueMap = boost::container::flat_map<std::string, std::string>;
+//add this type for negative filter support - ensure v2 structs encoded as (key, (value, is_negative)) 
+using KeyValueTypePairMap = boost::container::flat_map<std::string, std::pair<std::string,bool>>;
 using KeyMultiValueMap = std::multimap<std::string, std::string>;
 
 struct rgw_s3_key_value_filter {
-  KeyValueMap kv;
-
+  // key -> {value, is_negative} ; Type - IN => is_negative = false; Type - OUT => is_negative = true
+  KeyValueTypePairMap kvt; 
   bool has_content() const;
 
   void dump(Formatter *f) const;
   bool decode_xml(XMLObj *obj);
   void dump_xml(Formatter *f) const;
 
   void encode(bufferlist& bl) const {
-    ENCODE_START(1, 1, bl);
-      encode(kv, bl);
+    KeyValueMap kv; 
+    for(const auto& it: kvt){
+      if(it.second.second) {
+        kv.emplace(it.first, it.second.first);
+      }
+    }
+    ENCODE_START(2, 1, bl);
+    encode(kv, bl);
+    encode(kvt, bl);
     ENCODE_FINISH(bl);
   }
+
   void decode(bufferlist::const_iterator& bl) {
-    DECODE_START(1, bl);
+    KeyValueMap kv;
+    DECODE_START(2, bl);
       decode(kv, bl);
+      if(struct_v >= 2) {
+        decode(kvt, bl);
+      }
+      else {   
+        // convert kv to kvt for correct matching based on type for v1 structs
+        for (const auto& it : kv) {
+          kvt.emplace(it.first, std::make_pair(it.second, true));
```

Comment:
needs to be kvt.emplace(it.first, std::make_pair(it.second, false)); 
as defined on line 59.